File: leetcode/1396-design-underground-system/solution.cpp
#include <string>
#include <unordered_map>
using namespace std;

class UndergroundSystem {
public:
    UndergroundSystem() {}

    void checkIn(int id, string stationName, int t) {
        checkIns[id] = {stationName, t};
    }

    void checkOut(int id, string stationName, int t) {
        auto& in = checkIns[id];
        string key = in.first + "->" + stationName;
        auto& stats = trips[key];
        stats.first += t - in.second;
        stats.second += 1;
        checkIns.erase(id);
    }

    double getAverageTime(string startStation, string endStation) {
        auto& stats = trips[startStation + "->" + endStation];
        return (double)stats.first / stats.second;
    }

private:
    unordered_map<int, pair<string, int>> checkIns;      // id -> (station, t)
    unordered_map<string, pair<int, int>> trips;          // "start->end" -> (totalTime, count)
};

/**
 * Your UndergroundSystem object will be instantiated and called as such:
 * UndergroundSystem* obj = new UndergroundSystem();
 * obj->checkIn(id,stationName,t);
 * obj->checkOut(id,stationName,t);
 * double param_3 = obj->getAverageTime(startStation,endStation);
 */

File: leetcode/1396-design-underground-system/solution.py
class UndergroundSystem:
    def __init__(self):
        self.check_ins = {}   # id -> (station, t)
        self.trips = {}       # (start, end) -> [total_time, count]

    def checkIn(self, id: int, stationName: str, t: int) -> None:
        self.check_ins[id] = (stationName, t)

    def checkOut(self, id: int, stationName: str, t: int) -> None:
        start_station, start_t = self.check_ins.pop(id)
        key = (start_station, stationName)
        total_time, count = self.trips.get(key, (0, 0))
        self.trips[key] = (total_time + (t - start_t), count + 1)

    def getAverageTime(self, startStation: str, endStation: str) -> float:
        total_time, count = self.trips[(startStation, endStation)]
        return total_time / count


# Your UndergroundSystem object will be instantiated and called as such:
# obj = UndergroundSystem()
# obj.checkIn(id, stationName, t)
# obj.checkOut(id, stationName, t)
# param_3 = obj.getAverageTime(startStation, endStation)
